react/src/lib/components/WebvizDialog/WebvizDialog.tsx

```diff
@@ -0,0 +1,732 @@
+        } else {
+            node = document.getElementById(
+                "WebvizDialog__root"
+            ) as HTMLDivElement;
+        }
+
+        const placeholder = document.createElement("div");
+        setPlaceholderClassNameAndZIndex(placeholder, props.modal || false);
+        node.appendChild(placeholder);
+        setPlaceholderDiv(placeholder);
+
+        return () => {
+            if (placeholder) {
+                placeholder.parentElement?.removeChild(placeholder);
+            }
+            const dialogRoot = document.getElementById("WebvizDialog__root");
+            if (dialogRoot) {
+                const dialogNodes =
+                    dialogRoot.getElementsByClassName("WebvizDialog");
+                if (
+                    (dialogNodes.length == 1 &&
+                        dialogNodes[0] === dialogRef.current) ||
+                    dialogNodes.length === 0
+                ) {
+                    document.body.removeChild(dialogRoot);
+                }
+            }
+        };
+    }, []);
+
+    React.useLayoutEffect(() => {
+        if (props.open === open) {
+            return;
+        }
+
+        setOpen(props.open || false);
+        if (props.open) {
+            handleSetActive();
+        }
+    }, [props.open]);
+
+    const handleClose = React.useCallback(
+        (reason: DialogCloseReason) => {
+            if (reason == DialogCloseReason.BACKDROP_CLICK && !props.modal) {
+                return;
+            }
+            setOpen(false);
+            props.setProps({ open: false });
+        },
+        [props.modal, props.setProps]
+    );
+
+    React.useLayoutEffect(() => {
+        if (
+            !wrapperRef.current ||
+            wrapperRef.current.style.display !== "block" ||
+            !dialogRef.current
+        ) {
+            return;
+        }
+
+        let initialWidth = initialDialogWidth || 0;
+        const dialogHeight = dialogRef.current.getBoundingClientRect().height;
+        if (initialDialogWidth === null) {
+            initialWidth = dialogRef.current.getBoundingClientRect().width;
+            setInitialDialogWidth(initialWidth);
+        }
+
+        const top = Math.max(
+            margin.top,
+            Math.round(window.innerHeight / 2 - dialogHeight / 2)
+        );
+        setDialogTop(top);
+        if (window.innerWidth < margin.left + initialWidth + margin.right) {
+            const adjustedWidth = Math.max(
+                props.minWidth || 0,
+                Math.min(
+                    initialWidth,
+                    window.innerWidth - margin.left - margin.right
+                )
+            );
+            setDialogWidth(adjustedWidth);
+            setDialogLeft(margin.left);
+        } else {
+            const adjustedLeft = Math.round(
+                window.innerWidth / 2 - initialWidth / 2
+            );
+            setDialogWidth(initialWidth);
+            setDialogLeft(adjustedLeft);
```

Comment:
Could be advantageous to have a single state for all the dialog's client rect?

Reply:
Create `DialogRectData` object for state. Holding info regarding `top`, `left` and `width` for the dialog. Resize methods updated to calculate values and return, rather than setting individual states directly.